syntax = "proto3";

package dfs_service;

message Empty {}

message FileName {
    string name = 1;
}

message FileChunk {
    string filename = 1;
    bytes data = 2;
}

message FileStatus {
    string filename = 1;
    int32 size = 2;
    int32 mtime = 3;
    int32 ctime = 4;
}

message FileInfo {
    string name = 1;
    int32 mtime = 2;
}

message FileList {
    repeated FileInfo files = 1;
}

service DFSService {
    rpc Store(stream FileChunk) returns (FileStatus) {}
    rpc Fetch(FileName) returns (stream FileChunk) {}
    rpc Delete(FileName) returns (FileStatus) {}
    rpc List(Empty) returns (FileList) {}
    rpc Stat(FileName) returns (FileStatus) {}
}
